File: frontend-space/projects/parkpe/src/app/core/utils/transaction-id.ts
/**
 * Client-side ID matching backend `portal.utils.transaction_id.generate_transaction_id`.
 * Format (20 chars): T + YYYYMMDDHHMMSS (UTC) + 4 (ms-based) + 1 random alnum.
 * Example: T2026032007124729791
 */
const RAND = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ';

export function generateTransactionId(prefix = 'T'): string {
  const now = new Date();
  const p = (n: number, w: number) => String(n).padStart(w, '0');
  const ts = `${now.getUTCFullYear()}${p(now.getUTCMonth() + 1, 2)}${p(now.getUTCDate(), 2)}${p(
    now.getUTCHours(),
    2
  )}${p(now.getUTCMinutes(), 2)}${p(now.getUTCSeconds(), 2)}`;
  const micro4 = p(now.getMilliseconds() * 1000, 6).slice(0, 4);
  const rand1 = RAND[Math.floor(Math.random() * RAND.length)] ?? '0';
  const tid = `${prefix}${ts}${micro4}${rand1}`;
  return tid.length > 20 ? tid.slice(0, 20) : tid;
}

File: backend/portal/utils/transaction_id.py
"""
Central transaction / order / reference ID generator (ParkPe & portal).

Format (fixed 20 chars): T + YYYYMMDDHHMMSS + MMMM + R
  - T: prefix (default)
  - YYYYMMDDHHMMSS: UTC timestamp (14)
  - MMMM: first 4 digits of microsecond
  - R: one random char (0-9 or A-Z)

Example: T2026032007124729791
"""
from __future__ import annotations

import random
import string
from datetime import datetime


_RAND_CHARS = string.digits + string.ascii_uppercase


def generate_transaction_id(prefix: str = "T") -> str:
    """
    Generate a fixed-length 20-char transaction ID.

    1 (prefix) + 14 (datetime) + 4 (microsecond slice) + 1 (random char) = 20
    """
    now = datetime.utcnow()
    ts = now.strftime("%Y%m%d%H%M%S")
    micro4 = f"{now.microsecond:06d}"[:4]
    rand1 = random.choice(_RAND_CHARS)
    tid = f"{prefix}{ts}{micro4}{rand1}"
    # Safety guard for contract: always 20 chars.
    return tid[:20] if len(tid) > 20 else tid.ljust(20, "0")

